bofire/benchmarks/aspen_benchmark.py

```diff
@@ -0,0 +1,180 @@
+                )
+                logger.exception(log_string)
+                raise SystemExit(log_string)
+
+    # Start Aspen
+    def start_aspen(self):
+        """Starts Aspen plus and opens desired simulation file.
+
+        Raises:
+            SystemExit: In case it is not possible to start Aspen plus.
+        """
+        logger.info("Starting Aspen plus")
+        # Aspen should be accessible from every function as a global variable.
+        global aspen
+        try:
+            aspen = win32.Dispatch("Apwn.Document")
+            aspen.InitFromFile2(os.path.abspath(self.filename))
+            self.aspen_is_running = True
+        except OSError:
+            log_string = "Unable to start Aspen plus."
+            logger.exception(log_string)
+            raise SystemExit(log_string)
+
+    # Run simulation in Aspen
+    def _f(self, candidates: pd.DataFrame) -> pd.DataFrame:
+        """Function evaluation of the Aspen plus benchmark. Passes input values to Aspen, runs
+        a simulation and reads the output values from Aspen. Checks for errors/warnings from Aspen.
+
+        Args:
+            candidates (pd.DataFrame): Input values that need to be passed to Aspen.
+            Function can take either one input vector as a row in the input dataframe or multiple input vectors.
+
+        Returns:
+            pd.DataFrame: Output values from Aspen. The dataframe includes valid_(variable_name) columns for
+            each output variable when the simulation went successful.
+        """
+
+        # Only start Aspen, when it is not already blocking.
+        if self.aspen_is_running is False:
+            self.start_aspen()
+
+        # TODO: Initiate dataframe more memory efficient without interfering with the pd.concat method.
+        Y = pd.DataFrame()
+        # Iterate through dataframe rows to retrieve multiple input vectors. Running seperate simulations for each.
+        for index, row in candidates.iterrows():
+            logger.info("Writing inputs into Aspen")
+            # Write input variables corresping to columns into aspen according to predefined paths.
+            for key, path in zip(self.keys[0], self.paths[0]):
+                try:
+                    aspen.Tree.FindNode(path.get(key)).Value = row[key]
+                except ConnectionAbortedError:
+                    logger.exception("Not able to write " + key + " into Aspen.")
+
+            # Reset Aspen simulation
+            aspen.Reinit()
+            # Start new Aspen simulation
+            logger.info("Aspen simulation run " + str(index))
+            aspen.Engine.Run2()
+            logger.info("Simulation done.")
+
+            # Retrieve outputs from Aspen and write into data frame
+            logger.info("Retrieving outputs from Aspen.")
+            for key, path in zip(self.keys[1], self.paths[1]):
+                try:
+                    Y.at[index, key] = aspen.Tree.FindNode(path.get(key)).Value
```

Comment:
write the data first into a list/dictionary and return them in the end as dataframe